File: ZTMobile/ZTMobile/MainActivity.cs
﻿using System;
using Android.App;
using Android.Content;
using Android.Runtime;
using Android.Views;
using Android.Widget;
using Android.OS;
using System.Threading;
using Android.Support.V7.App;
using SupportToolbar = Android.Support.V7.Widget.Toolbar;
using Android.Support.V4.Widget;
using System.Collections.Generic;
using ZTMobile.Fragments;
using SupportFragment = Android.Support.V4.App.Fragment;
using Android.Gms.Maps.Model;
using Android.Gms.Maps;
using Android.Graphics;
using System.IO;
using Android.Views.Animations;
using System.Threading.Tasks;
using Android;
using Android.Content.PM;
using Android.Support.V4.App;
using Plugin.Permissions;
using Android.Support.V4.Content;

namespace ZTMobile
{
    [Activity(Label = "ZTMobile", MainLauncher = true, Icon = "@drawable/ic_launcher", Theme = "@style/MainTheme", ScreenOrientation = Android.Content.PM.ScreenOrientation.Portrait)]
    public class MainActivity : ActionBarActivity, ActivityCompat.IOnRequestPermissionsResultCallback
    {
        private SupportToolbar toolbar;
        private ActionBarDrawerToggle drawerToggle;
        private DrawerLayout drawerLayout;
        private ListView leftDrawer;
        private SupportFragment currentFragment;
        private TrackingScreen trackingScreenFragment;
        private LoginScreen loginScreenFragment;
        private MapScreen mapScreenFragment;
        private AccountScreen accountScreenFragment;
        private RankingScreen rankingScreenFragment;
        private DateTime lastBackButtonClickTime;
        private static bool pointOnActionBarLoaded = false;

        List<String> leftMenuItems = new List<String>();
        ArrayAdapter leftMenuArrayAdapter;

        protected override void OnCreate(Bundle bundle)
        {
            base.OnCreate(bundle);

            // Set our view from the "main" layout resource
            SetContentView(Resource.Layout.Main);

            lastBackButtonClickTime = DateTime.Now;

            CheckLocationPermission();
        }

        private void Initialize()
        {
            toolbar = FindViewById<SupportToolbar>(Resource.Id.toolbar);
            drawerLayout = FindViewById<DrawerLayout>(Resource.Id.drawer_layout);
            leftDrawer = FindViewById<ListView>(Resource.Id.left_drawer);

            leftMenuItems.Add("Śledzenie");
            leftMenuItems.Add("Konto");
            leftMenuItems.Add("Ranking");
            leftMenuItems.Add("Mapa");
            leftMenuArrayAdapter = new ArrayAdapter(this, Android.Resource.Layout.SimpleListItem1, leftMenuItems);
            leftDrawer.Adapter = leftMenuArrayAdapter;

            SetSupportActionBar(toolbar);

            trackingScreenFragment = new TrackingScreen();
            loginScreenFragment = new LoginScreen();
            mapScreenFragment = new MapScreen();
            accountScreenFragment = new AccountScreen();
            rankingScreenFragment = new RankingScreen();

            var transaction = SupportFragmentManager.BeginTransaction();
            transaction.Add(Resource.Id.fragmentContainer, accountScreenFragment, "Account Screen");
            transaction.Hide(accountScreenFragment);
            transaction.Add(Resource.Id.fragmentContainer, mapScreenFragment, "Map Screen");
            transaction.Hide(mapScreenFragment);
            transaction.Add(Resource.Id.fragmentContainer, rankingScreenFragment, "Ranking Screen");
            transaction.Hide(rankingScreenFragment);
            transaction.Add(Resource.Id.fragmentContainer, loginScreenFragment, "Login Screen");
            transaction.Hide(loginScreenFragment);
            transaction.Add(Resource.Id.fragmentContainer, trackingScreenFragment, "Tracking Screen");
            transaction.SetCustomAnimations(Android.Resource.Animation.FadeOut, Android.Resource.Animation.FadeIn);
            transaction.CommitAllowingStateLoss();

            currentFragment = trackingScreenFragment;

            drawerToggle = new ActionBarDrawerToggle(this, drawerLayout, Resource.String.openDrawer, Resource.String.closeDrawer);

            drawerLayout.SetDrawerListener(drawerToggle);
            SupportActionBar.SetHomeButtonEnabled(true);
            SupportActionBar.SetDisplayHomeAsUpEnabled(true);
            SupportActionBar.SetDisplayShowTitleEnabled(true);
            drawerToggle.SyncState();

            leftDrawer.ItemClick += LeftDrawer_ItemClick;
            loginScreenFragment.LoggedInSuccessfully += LoginScreenFragment_LoggedInSuccessfully;
            accountScreenFragment.LoggedOutSuccessfully += AccountScreenFragment_LoggedOutSuccessfully;
            accountScreenFragment.ActivityCreatedSuccessfully += AccountScreenFragment_ActivityCreatedSuccessfully;
            FunctionsAndGlobals.sendingFileCompleted += FunctionsAndGlobals_sendingFileCompleted;
        }

        private void FunctionsAndGlobals_sendingFileCompleted()
        {
            this.RunOnUiThread(() => { Toast.MakeText(this, Resource.String.fileSentMessage, ToastLength.Short).Show(); });
        }

        private void AccountScreenFragment_ActivityCreatedSuccessfully()
        {
            Thread thread = new Thread(() =>
                {
                    if (loginScreenFragment.LoginFromFile())
                    {
                        LoginScreenFragment_LoggedInSuccessfully();
                    }
                }
            );
            thread.Start();
        }

        private void PointsValueOnActionBar_TextChanged(object sender, Android.Text.TextChangedEventArgs e)
        {
            FunctionsAndGlobals.pointsValueOnActionBar.StartAnimation(AnimationUtils.LoadAnimation(this, Resource.Animation.abc_fade_in));

            if (FunctionsAndGlobals.pointsValueOnActionBar.Text != "")
                this.RunOnUiThread(() => { accountScreenFragment.ChangeVisiblePoints(Int32.Parse(FunctionsAndGlobals.pointsValueOnActionBar.Text)); });
        }

        private void LoginScreenFragment_LoggedInSuccessfully()
        {
            this.RunOnUiThread(() => { accountScreenFragment.ChangeVisibleUserName(FunctionsAndGlobals.userName); });

            if (currentFragment == loginScreenFragment)
            {
                ShowFragment(accountScreenFragment);
            }

            GetAndSetUserPoints();

            this.RunOnUiThread(() => { accountScreenFragment.SetProgressBarVisibilityState(true); });
            Bitmap bitmapImage = FunctionsAndGlobals.GetUserPhotoFromDatabase(FunctionsAndGlobals.userName);

            if (bitmapImage != null)
            {
                this.RunOnUiThread(() => { accountScreenFragment.ChangeVisibleUserPhoto(bitmapImage); });
            }
            this.RunOnUiThread(() => { accountScreenFragment.SetProgressBarVisibilityState(false); });
        }

        private void AccountScreenFragment_LoggedOutSuccessfully()
        {
            Thread thread = new Thread(() => ClearActionBar());
            thread.Start();

            if (currentFragment == accountScreenFragment)
            {
                ShowFragment(loginScreenFragment);
            }
        }

        public void ClearActionBar()
        {
            this.RunOnUiThread(() => { FunctionsAndGlobals.pointsTextOnActionBar.Text = ""; });
            this.RunOnUiThread(() => { FunctionsAndGlobals.pointsValueOnActionBar.Text = ""; });
        }

        public void GetAndSetUserPoints()
        {
            int result = -1;
            while (result == -1)
            {
                result = FunctionsAndGlobals.GetUserPointsFromDatabase(FunctionsAndGlobals.userName);
            }

            FunctionsAndGlobals.userPoints = result;

            while (!pointOnActionBarLoaded) { }

            this.RunOnUiThread(() => { FunctionsAndGlobals.pointsTextOnActionBar.Text = "Punkty:"; });
            this.RunOnUiThread(() => { FunctionsAndGlobals.pointsValueOnActionBar.Text = FunctionsAndGlobals.userPoints.ToString(); });

            this.RunOnUiThread(() => { accountScreenFragment.ChangeVisiblePoints(FunctionsAndGlobals.userPoints); });
        }

        private void LeftDrawer_ItemClick(object sender, AdapterView.ItemClickEventArgs e)
        {
            //Tracking Screen
            if (e.Position == 0)
            {
                ShowFragment(trackingScreenFragment);
            }
            //Login Screen
            if (e.Position == 1)
            {
                if (FunctionsAndGlobals.isUserLoggedIn == true)
                {
                    ShowFragment(accountScreenFragment);
                }
                else
                {
                    ShowFragment(loginScreenFragment);
                }
            }
            //Ranking Screen
            if (e.Position == 2)
            {
                ShowFragment(rankingScreenFragment);
            }
            //Map Screen
            if (e.Position == 3)
            {
                if (FunctionsAndGlobals.isTrackingEnabled == true)
                {
                    LatLng latLng = new LatLng(FunctionsAndGlobals.googleMap.MyLocation.Latitude, FunctionsAndGlobals.googleMap.MyLocation.Longitude);
                    CameraUpdate cameraUpdate = CameraUpdateFactory.NewLatLngZoom(latLng, 16);
                    FunctionsAndGlobals.googleMap.AnimateCamera(cameraUpdate);
                }

                ShowFragment(mapScreenFragment);
            }

            drawerLayout.CloseDrawer(Android.Support.V4.View.GravityCompat.Start);
        }

        public override bool OnOptionsItemSelected(IMenuItem item)
        {
            drawerToggle.OnOptionsItemSelected(item);
            return base.OnOptionsItemSelected(item);
        }

        public override bool OnCreateOptionsMenu(IMenu menu)
        {
            TextView pointsTextOnActionBar = new TextView(this);
            pointsTextOnActionBar.SetTextColor(Color.White);
            pointsTextOnActionBar.SetPadding(0, 0, 20, 0);
            pointsTextOnActionBar.SetTypeface(null, TypefaceStyle.Bold);
            pointsTextOnActionBar.SetTextSize(Android.Util.ComplexUnitType.Dip, 21);
            menu.Add(0, 1, 1, Resource.String.error).SetActionView(pointsTextOnActionBar).SetShowAsAction(ShowAsAction.Always);
            pointsTextOnActionBar.Text = "";
            FunctionsAndGlobals.pointsTextOnActionBar = pointsTextOnActionBar;

            TextView pointsValueOnActionBar = new TextView(this);
            pointsValueOnActionBar.SetTextColor(Color.White);
            pointsValueOnActionBar.SetPadding(0, 0, 20, 0);
            pointsValueOnActionBar.SetTypeface(null, TypefaceStyle.Bold);
            pointsValueOnActionBar.SetTextSize(Android.Util.ComplexUnitType.Dip, 21);
            menu.Add(0, 1, 2, Resource.String.error).SetActionView(pointsValueOnActionBar).SetShowAsAction(ShowAsAction.Always);
            pointsValueOnActionBar.Text = "";
            FunctionsAndGlobals.pointsValueOnActionBar = pointsValueOnActionBar;

            FunctionsAndGlobals.pointsValueOnActionBar.TextChanged += PointsValueOnActionBar_TextChanged;

            pointOnActionBarLoaded = true;
            return base.OnCreateOptionsMenu(menu);
        }

        public override void OnBackPressed()
        {
            DateTime pressTime = DateTime.Now;
            if ((pressTime - lastBackButtonClickTime).TotalMilliseconds <= FunctionsAndGlobals.doubleBackButtonClickInterval_ms)
            {
                Java.Lang.JavaSystem.Exit(0);
            }

            Toast.MakeText(this, Resource.String.exitMessage, ToastLength.Short).Show();
            lastBackButtonClickTime = pressTime;
        }

        private void ShowFragment(SupportFragment fragment)
        {
            var transaction = SupportFragmentManager.BeginTransaction();
            transaction.Hide(currentFragment);
            transaction.SetCustomAnimations(Android.Resource.Animation.SlideInLeft, Android.Resource.Animation.SlideOutRight);
            transaction.Show(fragment);
            transaction.AddToBackStack(null);
            transaction.Commit();

            currentFragment = fragment;
        }

        private void CheckLocationPermission()
        {
            if (ContextCompat.CheckSelfPermission(this, Manifest.Permission.AccessFineLocation) == (int)Android.Content.PM.Permission.Granted)
            {
                Initialize();
                return;
            }

            ActivityCompat.RequestPermissions(this, new String[] { Manifest.Permission.AccessFineLocation }, 1);
        }

        public override async void OnRequestPermissionsResult(int requestCode, string[] permissions, [GeneratedEnum] Permission[] grantResults)
        {
            switch (requestCode)
            {
                case 1:
                    {
                        // If request is cancelled, the result arrays are empty.
                        if (grantResults.Length > 0 && grantResults[0] == Permission.Granted)
                        {
                            Initialize();
                            return;
                        }
                        else
                        {
                            Java.Lang.JavaSystem.Exit(0);
                        }
                        break;
                    }
            }
        }
    }
}

File: ZTMobile/app/src/main/java/example/com/ztmobile/MainActivity.java
package example.com.ztmobile;

import android.content.Intent;
import android.support.annotation.IdRes;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.app.ActionBarActivity;
import android.support.v7.widget.Toolbar;
import android.os.Bundle;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.ListView;
import android.widget.Toast;
import android.os.Handler;

import com.google.android.gms.maps.CameraUpdate;
import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.model.LatLng;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity {

    private Toolbar toolbar;
    private ActionBarDrawerToggle drawerToggle;
    private DrawerLayout drawerLayout;
    private ListView leftDrawer;
    private Fragment currentFragment;
    private boolean isBackKeyPressedTwice;
    private TrackingScreen trackingScreenFragment;
    private LoginScreen loginScreenFragment;
    private MapScreen mapScreenFragment;
    //private AccountScreen accountScreenFragment;
    //private DateTime lastBackButtonClickTime;

    List<String> leftMenuItems = new ArrayList<String>();
    ArrayAdapter leftMenuArrayAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        isBackKeyPressedTwice = false;
        toolbar = (Toolbar) findViewById(R.id.toolbar1);
        drawerLayout = (DrawerLayout) findViewById(R.id.drawer_layout);
        leftDrawer = (ListView) findViewById(R.id.left_drawer);

        leftMenuItems.add("Śledzenie");
        leftMenuItems.add("Konto");
        leftMenuItems.add("Mapa");
        leftMenuArrayAdapter = new ArrayAdapter(this, android.R.layout.simple_list_item_1, leftMenuItems);
        leftDrawer.setAdapter(leftMenuArrayAdapter);
        leftDrawer.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position,
                                    long id) {
                switch (position)
                {
                    case 0:
                        ShowFragment(trackingScreenFragment);
                        break;

                    case 1:
                        ShowFragment(loginScreenFragment);
                        break;

                    case 2:
                        if (FunctionsAndGlobals.isTrackingEnabled == true)
                        {
                            LatLng latLng = new LatLng(FunctionsAndGlobals.googleMap.getMyLocation().getLatitude(), FunctionsAndGlobals.googleMap.getMyLocation().getLongitude());
                            CameraUpdate cameraUpdate = CameraUpdateFactory.newLatLngZoom(latLng, 16);
                            FunctionsAndGlobals.googleMap.animateCamera(cameraUpdate);
                        }
                        ShowFragment(mapScreenFragment);
                        break;

                    default:
                        break;
                }

                drawerLayout.closeDrawer(GravityCompat.START);
            }
        });

        setSupportActionBar(toolbar);

        mapScreenFragment = new MapScreen();
        trackingScreenFragment = new TrackingScreen();
        loginScreenFragment = new LoginScreen();

        FragmentTransaction transaction = getSupportFragmentManager().beginTransaction();
        //transaction.add(Resource.Id.fragmentContainer, accountScreenFragment, "Account Screen");
        //transaction.hide(accountScreenFragment);
        transaction.add(R.id.fragmentContainer, mapScreenFragment, "Map Screen");
        transaction.hide(mapScreenFragment);
        //.add(Resource.Id.fragmentContainer, loginScreenFragment, "Login Screen");
        //transaction.hide(loginScreenFragment);
        transaction.add(R.id.fragmentContainer, trackingScreenFragment, "Tracking Screen");
        transaction.commit();

        currentFragment = trackingScreenFragment;

        drawerToggle = new ActionBarDrawerToggle(MainActivity.this, drawerLayout, R.string.openDrawer, R.string.closeDrawer);

        drawerLayout.setDrawerListener(drawerToggle);
        getSupportActionBar().setHomeButtonEnabled(true);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setDisplayShowTitleEnabled(true);
        drawerToggle.syncState();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Pass the event to ActionBarDrawerToggle, if it returns
        // true, then it has handled the app icon touch event
        if (drawerToggle.onOptionsItemSelected(item)) {
            return true;
        }
        // Handle your other action bar items...

        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onBackPressed()
    {
        if (isBackKeyPressedTwice == true)
        {
            Intent intent = new Intent(Intent.ACTION_MAIN);
            intent.addCategory(Intent.CATEGORY_HOME);
            intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
            startActivity(intent);
            finish();
            System.exit(0);
        }

        Toast.makeText(MainActivity.this, R.string.exitMessage, Toast.LENGTH_SHORT).show();

        new android.os.Handler().postDelayed(new Runnable() {
            @Override
            public void run() {
                isBackKeyPressedTwice = false;
            }
        }, FunctionsAndGlobals.doubleBackButtonClickInterval_ms);

        isBackKeyPressedTwice = true;
    }

    private void ShowFragment(Fragment fragment)
    {
        FragmentTransaction transaction = getSupportFragmentManager().beginTransaction();
        transaction.hide(currentFragment);
        transaction.show(fragment);
        transaction.addToBackStack(null);
        transaction.commit();

        currentFragment = fragment;
    }
}
